use a die directly in the @INC hook

haarg opened the pull request:
Using a die directly in the `@INC` hook matches perl's behavior for a
missing module closer than using a fake module that dies.  It is also
simpler and more compatible.

On perl 5.8, with a die inside a module, the `%INC` entry will be left
behind.  Later attempts to load the module will then appear to succeed.
Putting the die directly will leave the `%INC` entry non-existent, just
like if the file does not exist.

This also avoids PerlIO, so it works on perl 5.6 without any extra work.

The error message was also updated to include ` in @INC`, so it is
closer to the error provided by perl.  This is expected by some modules
handling optional loads, such as Class::Load.

The behaviors are demonstrated in https://gist.github.com/haarg/c55c702c08fc59a8d61b1265f1c07353

haarg committed:
use a die directly in the @INC hook

Using a die directly in the @INC hook matches perl's behavior for a
missing module closer than using a fake module that dies.  It is also
simpler and more compatible.

On perl 5.8, with a die inside a module, the %INC entry will be left
behind.  Later attempts to load the module will then appear to succeed.
Putting the die directly will leave the %INC entry non-existent, just
like if the file does not exist.

This also avoids PerlIO, so it works on perl 5.6 without any extra work.

The error message was also updated to include " in @INC", so it is
closer to the error provided by perl.  This is expected by some modules
handling optional loads, such as Class::Load.

ap commented:
Bonus is that this probably also makes it work pre-5.8, for anyone who cares about that.

ap commented:
Also, one coderef per module seems rather wasteful. Is there a reason not to use just a single coderef?

``` perl
use lib sub { die "Can't locate $_[1] (hidden)\n" if grep { $_[1] eq $_ } qw( Cpanel/JSON/XS.pm ) };
```

And of course to exclude only one module one can shorten out the `grep`.

``` perl
use lib sub { die "Can't locate $_[1] (hidden)\n" if $_[1] eq 'Cpanel/JSON/XS.pm' };
```

In fact, this pattern is now so simple, it crosses the threshold from a snippet to keep on file for copy&pasting to an idea you can just memorise and type out. I would therefore advocate for putting that one in the docs first, and then mention the `grep` variant as a “btw, to hide multiple modules, obviously, …”.

haarg commented:
I've updated this to fix the rest of the documentation so it corresponds to this change.  I also changed the error message used to include `in @INC`, since some modules like Class::Load expect that.
